### ISSUE TITLE
[meta] Ship v2

### ISSUE THREAD

**Issue by mjackson** [rackt/history#172]
I think the only thing that's left to ship v2 is to remove all the deprecated/weird code and release. Any objections?

/cc @rackt/routing

**Comment by taion** [comment]
These two need to be fixed: https://github.com/rackt/history/blob/master/modules/useQueries.js#L107-L113

And dropping `createLocation` like you said before.

**Comment by taion** [comment]
Oh, and https://github.com/rackt/history/issues/163 - if we want to change the default for `createHashHistory`.

**Comment by taion** [comment]
We're also in a bit of an odd spot right now w/r/t the deprecations, because we haven't updated React Router's internals yet to switch to the new APIs (e.g. on `<Link>`).

Especially if we think we'll ship history v2.x before we ship React Router v1.1.x, then we might want to put out a stopgap release of React Router that doesn't change its APIs, but updates to use the new history APIs.

**Comment by timdorr** [comment]
We can just delay the deprecations until after we land and release those other issues you've mentioned.

**Comment by taion** [comment]
:+1:

**Comment by taion** [comment]
Or unwind them anyway.

**Comment by mjackson** [comment]
I'd like to get history v2 out the door ASAP just so I can check something off my TODO list. Too many loose ends makes it easy for me to forget something :)

I don't actually consider #163 a breaking change. Everything should work as before after that change, including URLs with the hash. Also, I'm not 100% sure what the implementation is going to look like there, so I'm happy to move forward with history v2 without it.

Now that #173 is merged, I think we're ready to cut a 1.15 release. v2 after that will just be removing a bunch of deprecated stuff.

**Comment by taion** [comment]
What do you think about shipping a React Router release that uses the new API, minus the Router-side API changes? I definitely want these API cleanups, but I also want our users to be able to use them. As-is, Router users won't be able to use history v2 at all until we ship something semi-updated (mostly just https://github.com/rackt/react-router/pull/2659) on the Router side.

**Comment by mjackson** [comment]
Agreed. Let's ship a router 1.1 release that uses history v2. We can put the changes that are specific to https://github.com/rackt/react-router/issues/2646 in a separate branch to be merged back into master after the 1.1 release.

**Comment by timdorr** [comment]
I'm in agreement about a v1.15.0 release here. We should use that version in router 1.1, which will remove usage of any deprecated APIs. Then we can go 2.0 on router and history in lockstep, easing the transition path, since they won't have to update things twice.

**Comment by mjackson** [comment]
ok, so are we ready for a 1.15 release? Looks to me like we are.

**Comment by timdorr** [comment]
Yep, you can cut that right now.

**Comment by taion** [comment]
With the `<Link>` PR merged, does the router have any dependencies on the old APIs? At that point it should work with 2.x as well. I think we can support both history 1.15.x and 2.x by using just the 2.x API.

A little unfortunate to have to move some code out of the way for a bit. I didn't realize we'd get these changes on the history side ready so quickly. That's a very good problem to have though! (:

**Comment by timdorr** [comment]
[Yep](https://github.com/rackt/react-router/search?q=pushstate). It's mostly minor stuff. I was planning on getting all of that alongside updating things to use `this.context.router`, but I can move up that effort to remove any deprecation warnings.

**Comment by taion** [comment]
Yeah, I think we want to move the `context.router` stuff onto a branch, and just get rid of the deprecated API use. The changes actually shouldn't collide at all. Then we can just mark the peer dependency as `^1.15.0 || 2.x`.

**Comment by mjackson** [comment]
OK. I'll cut a 1.15 history release and we can remove the last remaining bits from the router that use pushState/replaceState.

**Comment by mjackson** [comment]
I'll dig in here and try to get a PR up by EOD that removes all deprecations for v2. I'm sure I'll miss a few things, but hopefully you guys will be able to catch me :)

If either of you would like to take a stab at a PR on the router that makes the changes we discussed above that would be great.

**Comment by mjackson** [comment]
ok, now that we've shipped 1.16 that should fix everyone's builds and stop the flow of issues in the router. If, at this point, we ship a version 1.17 with the warnings back in, we'll just break everyone's build again.

So, I propose that we ship what we currently have on master as history v2. It's full of deprecation warnings, but the recommended API is pretty much exactly what we want it to be and people won't get auto-upgraded to it. In a few weeks, we can ship history v3 that removes all the deprecated code. In the mean time, we can ship bug fixes and features on the 2.x line.

**Comment by timdorr** [comment]
Agreed. @ryanflorence's logic for doing a major version deprecation schedule on router would apply well here too.

I can get a 1.0.3 of router together so people get upgraded without confusion over peerDeps or warnings. Can you publish that?

**Comment by mjackson** [comment]
@timdorr yep, that'd be great. Thank you :)

**Comment by timdorr** [comment]
Ready in rackt/react-router#2702. It'll all be ready in the `1.0.x` branch.

**Comment by timdorr** [comment]
So, we've removed all the relevant deprecation warnings. Are we going to ship 2.0 with them added back in or just not ship a warning version?

**Comment by taion** [comment]
We need a version with both the new API and the deprecations to give users a good upgrade path. From Discord, it sounds like it'd be better to make that 2.0.0 rather than 1.18.0.

**Comment by ryanflorence** [comment]
Yes, 2.0 with warnings, then in no less than 3 months, release again w/o them.

**Comment by mjackson** [comment]
I'm planning on shipping 2.0 w warnings tonight. If anyone has any last-minute input, please speak up!

**Comment by taion** [comment]
What do you think of shipping this as an RC just in case we missed something? We caught a few things over the last few days in the Router that we missed before cutting the initial RC. It'd be awkward to have to do another major version bump or something just in case we need to change some API we didn't notice.

**Comment by mjackson** [comment]
Sure. I'll cut 2.0.0-rc1 tonight then. Thanks for the suggestion.

**Comment by mjackson** [comment]
ok, [2.0.0-rc1 is out](https://github.com/rackt/history/releases/tag/v2.0.0-rc1). Please kick the tires.

**Comment by taion** [comment]
Our dist-tags look a bit weird on npm:

```sh
jjia@jjia-mbp:~/vendor/react-router (upgrade-history-dep)$ npm dist-tag ls history
latest: 2.0.0-rc1

jjia@jjia-mbp:~/vendor/react-router (upgrade-history-dep)$ npm dist-tag ls react-router
beta: 2.0.0-rc4
latest: 2.0.0-rc4
```

Probably we should still have respectively 1.17.0 and 1.0.3 on `latest` with the RCs tagged `beta`?

**Comment by mjackson** [comment]
Good catch, @taion. Though I think it's a bit weird to use the `beta` tag instead of `rc`. I'll update.

**Comment by mjackson** [comment]
Just an update here: I'd like to use the `next` npm dist-tag across all our repos for the next upcoming release. It's also what React core uses:

```
$ npm dist-tag ls react
0.10.0-rc1: 0.10.0-rc1
0.11.0-rc1: 0.11.0-rc1
latest: 0.14.5
next: 0.14.0-rc1
```

**Comment by knowbody** [comment]
:+1: @mjackson 





On Wed, Jan 6, 2016 at 11:21 AM -0800, "Michael Jackson" <notifications@github.com> wrote:












Just an update here: I'd like to use the next npm dist-tag across all our repos for the next upcoming release. It's also what React core uses:

$ npm dist-tag ls react
0.10.0-rc1: 0.10.0-rc1
0.11.0-rc1: 0.11.0-rc1
latest: 0.14.5
next: 0.14.0-rc1

**Comment by timdorr** [comment]
James was doing the same thing on redux-simple-router too. I'm all for it!

**Comment by timdorr** [comment]
@mjackson Since we got #200 merged in, can we get a 2.0.0-rc3 cut? It fixes some things needed for router to go 2.0. 

We should be able to do a 2.0.0-rc6 on router just as a last-minute double-check for issues and be good to go 2.0 here and on router. :metal:

**Comment by mjackson** [comment]
Sorry for the silence everyone. Been traveling a lot lately. I'll cut 2.0.0-rc3 today.

**Comment by taion** [comment]
Great, thanks! Let me fix #

**Comment by taion** [comment]
Great, thanks! Let me fix https://github.com/rackt/history/issues/221 real quick.

**Comment by taion** [comment]
Sorry, finger slipped.

**Comment by mjackson** [comment]
Awesome, thanks @taion. Anything else you'd like to do before we cut version 2? We can always do minor/patch releases for new API and/or bugs afterward.

**Comment by taion** [comment]
Nope, this should be it – just want to get this in because it's a pretty trivial fix.

**Comment by taion** [comment]
Fix is on https://github.com/rackt/history/pull/222.

**Comment by mjackson** [comment]
OK, looks like we're good to go ahead and cut v2.

**Comment by taion** [comment]
:+1:

**Comment by taion** [comment]
@mjackson Can you fix the npm tags? This doesn't look right:

```
latest: 2.0.0-rc3
next: 2.0.0-rc1
rc: 2.0.0-rc1
```

**Comment by mjackson** [comment]
@taion Fixed dist-tags and cut v2.0.0 release today :tada:

Thanks everyone for your help!
